src/niri.rs

```diff
@@ -647,7 +662,60 @@ impl State {
         // build up (the 1 second frame callback timer will call this line).
         self.niri.advance_animations();
 
-        self.niri.redraw_queued_outputs(&mut self.backend);
+        let mut queued_outputs = self.niri.queued_outputs();
+        queued_outputs.sort_by_key(|(_, target_presentation_time, _)| *target_presentation_time);
+
+        let mut min_next_schedule: Option<Duration> = None;
+
+        for (output, target_presentation_time, refresh_interval) in queued_outputs {
+            let next_deadline = self.signal_commit_timing(&output, target_presentation_time);
+
+            if let Some(next_deadline) = next_deadline {
+                let refresh_interval =
+                    refresh_interval.unwrap_or_else(|| Duration::from_millis(16));
+
+                let steps = next_deadline
+                    .saturating_sub(target_presentation_time)
+                    .div_duration_f64(refresh_interval) as u32;
+
+                let next_schedule = refresh_interval
+                    .saturating_mul(steps)
+                    .saturating_add(target_presentation_time);
+
+                debug!(
+                    output = output.name(),
+                    ?target_presentation_time,
+                    ?next_deadline,
+                    ?next_schedule,
+                    ?refresh_interval,
+                    steps
+                );
+
+                min_next_schedule = min_next_schedule
+                    .filter(|min_next_schedule| *min_next_schedule <= next_schedule)
+                    .or(Some(next_schedule));
+            }
+
+            if self.niri.is_queued(&output) {
+                self.niri
+                    .redraw(&mut self.backend, &output, target_presentation_time);
+                self.signal_fifo(&output);
```

Comment:
Same question here about clearing commit blockers at the end here, after `layout.refresh()`. Though this one is trickier because I understand it has to happen after a redraw which itself has to happen at the end.